src/SecureRoute.js

```diff
@@ -15,19 +15,29 @@ import { useOktaAuth } from './OktaContext';
 import { Route, useRouteMatch } from 'react-router-dom';
 
 const SecureRoute = ( props ) => { 
-  const { authService, authState } = useOktaAuth();
+  const { oktaAuth, authState, _onAuthRequired } = useOktaAuth();
   const match = useRouteMatch(props);
 
   useEffect(() => {
+    const handleLogin = async () => {
+      oktaAuth.setOriginalUri();
+      const onAuthRequiredFn = props.onAuthRequired || _onAuthRequired;
+      if (onAuthRequiredFn) {
+        await onAuthRequiredFn(oktaAuth);
+      } else {
+        await oktaAuth.signInWithRedirect();
+      }
+    };
+
     // Only process logic if the route matches
     if (!match) {
       return;
     }
     // Start login if and only if app has decided it is not logged inn
     if(!authState.isAuthenticated && !authState.isPending) { 
-      authService.login();
+      handleLogin();
```

Comment:
Is there a problem with calling an `async` function within a `useEffect` block? What if the component renders again while we are awaiting the `onAuthRequired` function? Would it make sense to use something like state to prevent multiple calls?

Reply:
The isPending options should prevent multiple calls - at least, that was part of the role when it was in okta-react directly.

Reply:
The default `signInWithRedirect` will take us to Okta and so unload the entire application.
But `onAuthRequired` may never resolve before it cases actions that make `isAuthenticated` to change to true. Then, something else happens which causes `isAuthenticated` to become false, and it will call `onAuthRequired` again, possibly before the first call has returned.  It's an edge case, but imagine instead of a redirect flow we are doing an interact flow and we are waiting for a code from a mobile app. Meanwhile actions in other tabs cause the tokens to change which update auth state and toggle the value of `isAuthenticated` causing a re-render. Once we enter an auth flow and show the user a form, I don't think we'd want to re-render and erase the progress the user has made on that form, no matter what is happening under the hood.

Reply:
> Is there a problem with calling an async function within a useEffect block?

Yes, use `async` directly with `useEffect` will cause warning in console.

> Would it make sense to use something like state to prevent multiple calls?

Agree. We already implemented logic to prevent multiple calls for `signInWithRedirect` https://github.com/okta/okta-auth-js/blob/master/lib/browser/browser.ts#L264, the callback options should also be protected with similar logic.

Reply:
Protected with `pendingRef` in SecureRoute